### PR TITLE
feat!: support new algoliasearch and django versions

### PR DESCRIPTION
**Pull request by shortcuts**
## 🧭 What and Why

🎟 JIRA Ticket: https://algolia.atlassian.net/browse/DI-3186 https://algolia.atlassian.net/browse/DI-3187 https://algolia.atlassian.net/browse/DI-3188 https://algolia.atlassian.net/browse/DI-3189 https://algolia.atlassian.net/browse/DI-3191 https://algolia.atlassian.net/browse/DI-3247
https://algolia.atlassian.net/browse/DI-3190

closes https://github.com/algolia/algoliasearch-django/issues/333
closes https://github.com/algolia/algoliasearch-django/issues/249

### Changes included:

upgrade the package to support django v4 and v5, also upgrade the algoliasearch client

List of breaking changes:
- drop support for python <3.8
- drop support for django <4.0
- remove method `clear_index`, use `clear_objects` instead
- upgrade to [algoliasearch](https://pypi.org/project/algoliasearch/) +4.0

### CODE REVIEW

File: algoliasearch_django/models.py [hunk 1]
```diff
@@ -350,87 +368,87 @@ def update_records(self, qs, batch_size=1000, **kwargs):
...
         for elt in objectsIDs:
-            tmp['objectID'] = elt
+            tmp["objectID"] = elt
             batch.append(dict(tmp))
 
-            if len(batch) >= batch_size:
-                self.__index.partial_update_objects(batch)
-                batch = []
-
+        # TODO: pass batch_size to partial_update_objects
```
Comment (shortcuts) [inline algoliasearch_django/models.py:377]: ah yes I need to support that

File: requirements.txt [hunk 2]
```diff
@@ -1,9 +1,14 @@
-django>=1.7
-algoliasearch>=3.0,<4.0
+Django>=4.0
+# algoliasearch>=4.0,<5.0
+algoliasearch @ git+https://github.com/algolia/algoliasearch-client-python@main
```
Comment (shortcuts) [inline requirements.txt:3]: waiting for bfcm to end for this

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -1,22 +1,17 @@
-| Q                 | A
-| ----------------- | ----------
-| Bug fix?          | yes/no
-| New feature?      | yes/no    <!-- please update the /CHANGELOG.md file -->
-| BC breaks?        | no     
-| Related Issue     | Fix #...  <!-- will close issue automatically, if any -->
-| Need Doc update   | yes/no
+## 🧭 What and Why
 
+🎟 JIRA Ticket:
```
Comment (millotp) [inline .github/PULL_REQUEST_TEMPLATE.md:3]: maybe `Related Issue` was better, external contributor don't care about jira

**Review by millotp — COMMENTED** [review]
soooo nice, gg

File: tests/test_signal.py [hunk 4]
```diff
@@ -44,31 +43,26 @@ def test_save_signal(self):
...
-        )
+        mocked_delete_record.assert_has_calls([call(websites[0]), call(websites[1])])
 
     def test_update_records(self):
-        Website.objects.create(name='Algolia', url='https://www.algolia.com')
-        Website.objects.create(name='Google', url='https://www.google.com')
-        Website.objects.create(name='Facebook', url='https://www.facebook.com')
-        Website.objects.create(name='Facebook', url='https://www.facebook.fr')
-        Website.objects.create(name='Facebook', url='https://fb.com')
+        Website(name="Algolia", url="https://www.algolia.com", is_online=False)
```
[the remark this answers is not in the document]
Comment (shortcuts) [reply tests/test_signal.py:53]: I went with the default configuration just to have simpler maintenance logic but... I don't really mind

File: algoliasearch_django/registration.py [hunk 5]
```diff
@@ -30,17 +26,18 @@ def __init__(self, settings=SETTINGS):
...
 
-        self.__auto_indexing = settings.get('AUTO_INDEXING', True)
+        self.__auto_indexing = settings.get("AUTO_INDEXING", True)
         self.__settings = settings
 
         self.__registered_models = {}
-        self.client = SearchClient.create(app_id, api_key)
+        self.client = SearchClientSync(app_id, api_key)
+        self.client.add_user_agent("Algolia for Django", __version__)
+        self.client.add_user_agent("Django", __django__version__)
```
[the remark this answers is not in the document]
Comment (shortcuts) [reply algoliasearch_django/registration.py:40]: actually it's me that names it like that at the import level, just for consistency sake but it seems to be one of the support solution for django!

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 3 above)
Comment (shortcuts) [reply .github/PULL_REQUEST_TEMPLATE.md]: indeed!

**Review by morganleroi — APPROVED** [review]
Thanks for taking care of Django

**Pull request merged**